ingestion/ca/open_gov_canada.py

```diff
@@ -0,0 +1,183 @@
+from typing import Generator
+from dateutil.parser import parse as parse_date
+from ingestion.data_models import UpstreamDataset, PartialDataset, AltStr
+import httpx
+import lxml.html
+import json
+import csv
+import io
+import re
+from functools import reduce
+
+SITE_URL = "https://search.open.canada.ca/opendata/"
+CSV_URL = "https://open.canada.ca/data/dataset/c4c5c7f1-bfa6-4ff6-b4a0-c164cb2060f7/resource/312a65c5-d0bc-4445-8a24-95b2690cc62b/download/main.csv"
+RECORD_URL = "https://open.canada.ca/data/api/action/package_show?id={}"
+
+
+def make_request(url):
+    """
+    Make an HTTP request with logging & error checking.
+    """
+    resp = httpx.get(url, follow_redirects=True)
+    resp.raise_for_status()
+    return resp
+
+
+def list_datasets() -> Generator[PartialDataset, None, None]:
+    """
+    This method can either return a list or `yield` individual
+    items as they're found.
+
+    The version below demonstrates yielding items, but
+    """
+    data = make_request(CSV_URL)
+    for row in csv.DictReader(io.StringIO(data.text)):
+        # fill last_updated with date_modified if it exists or date_published if not
+        if row["date_modified"] == "":
+            last_updated = parse_date(row["date_published"])
+        else:
+            last_updated = parse_date(row["date_modified"])
+        yield PartialDataset(
+            url=RECORD_URL.format(row["id"]),
+            last_updated=last_updated,
+        )
+
+
+def get_dataset_details(pd: PartialDataset) -> UpstreamDataset:
+    # parse HTML from request
+    resp = make_request(pd.url)
+    root = lxml.html.fromstring(resp.content)
```

Comment:
Is the response that comes back from this URL JSON? If so, you can skip the lxml.html.fromstring step.